server/aws-lsp-codewhisperer/src/language-server/telemetryService.ts

```diff
@@ -0,0 +1,122 @@
+    }
+
+    public updateOptOutPreference(optOutPreference: OptOutPreference): void {
+        this.optOutPreference = optOutPreference
+    }
+
+    public updateEnableTelemetryEventsToDestination(enableTelemetryEventsToDestination: boolean): void {
+        this.enableTelemetryEventsToDestination = enableTelemetryEventsToDestination
+    }
+
+    private getSuggestionState(session: CodeWhispererSession): SuggestionState {
+        let suggestionState: SuggestionState
+        switch (session.getAggregatedUserTriggerDecision()) {
+            case 'Accept':
+                suggestionState = 'ACCEPT'
+                break
+            case 'Reject':
+                suggestionState = 'REJECT'
+                break
+            case 'Discard':
+                suggestionState = 'DISCARD'
+                break
+            default:
+                suggestionState = 'EMPTY'
+        }
+        return suggestionState
+    }
+
+    private isLoginInvalidForTelemetry(): boolean {
+        return this.credentialsType === 'iam' || (this.loginType === 'builderId' && this.optOutPreference === 'OPTOUT')
+    }
+
+    private invokeSendTelemetryEvent(event: any) {
+        const request: SendTelemetryEventRequest = {
+            telemetryEvent: {
+                userTriggerDecisionEvent: event,
+            },
+        }
+        if (this.userContext !== undefined) {
+            request.userContext = this.userContext
+        }
+        if (this.optOutPreference !== undefined) {
+            request.optOutPreference = this.optOutPreference
+        }
+        this.sendTelemetryEvent(request)
+    }
+
+    public emitUserTriggerDecision(session: CodeWhispererSession, timeSinceLastUserModification?: number) {
+        if (this.isLoginInvalidForTelemetry()) {
+            return
+        }
+        const completionSessionResult = session.completionSessionResult ?? {}
+        const acceptedSuggestion = session.suggestions.find(s => s.itemId === session.acceptedSuggestionId)
+        const generatedLines =
+            acceptedSuggestion === undefined || acceptedSuggestion.content.trim() === ''
+                ? 0
+                : acceptedSuggestion.content.split('\n').length
+        const referenceCount =
+            acceptedSuggestion === undefined
+                ? 0
+                : acceptedSuggestion.references && acceptedSuggestion.references.length > 0
+                  ? 1
+                  : 0
+
+        const event: UserTriggerDecisionEvent = {
+            sessionId: session.codewhispererSessionId || '',
+            requestId: session.responseContext?.requestId || '',
+            customizationArn: session.customizationArn === '' ? undefined : session.customizationArn,
+            programmingLanguage: {
+                languageName: getRuntimeLanguage(session.language),
+            },
+            completionType:
+                session.suggestions.length > 0 ? getCompletionType(session.suggestions[0]).toUpperCase() : '',
```

Comment:
Looks like completionType can't be empty string https://github.com/aws/language-servers/blob/4aee4d35769d1bfc79fbf3b3b8e5b7a3330a327a/server/aws-lsp-codewhisperer/src/client/token/bearer-token-service.json#L1090-L1093

VSCode defaults to `LINE`, guess we should do the same